### ISSUE TITLE
[Blazor] Avoid completely replacing the DOM when components are re-rendered after prerendering

### ISSUE THREAD

**Issue by javiercn** [dotnet/aspnetcore#42561]
Right now, after prerendering we replace the entire contents of the prerendered component with a fresh DOM copy created from scratch. This has several bad side-effects like:
* Videos are stopped and restarted
* Animations can be retriggered
* The page can flicker if CSS is injected as part of the head contents.
* LCP score suffers because (even though there is no significant visual change in most cases) the tools detect the interactive DOM insertion as the LCP element.

**Comment by Freshmintyy** [comment]
Yes yes yes. This is long overdue my guys 🙏😍😇

**Comment by aryehsilver** [comment]
Great to see this is finally being worked on. Let's hope it doesn't get pushed off any longer.

**Comment by fingers10** [comment]
You can refer [ilovedotnet](https://www.ilovedotnet.org) for this issue. The site is built from scratch using blazor wasm and the site also facing flicker issue when it gets loaded. The site has prerendering enabled during publish time in the pipeline. When the site loads the prerendered screen appears and when the blazor engine kicks in, there comes a flash and screen appears after that.

**Comment by kevon-vems** [comment]
Following

**Comment by oluatte** [comment]
Also interested in this. We're building a consumer facing web app & the flicker is very noticeable.

**Comment by rogihee** [comment]
This one hurts pretty badly with .NET 8 if you have SSR with any kind of interactivity somewhere

**Comment by andersme** [comment]
> This one hurts pretty badly with .NET 8 if you have SSR with any kind of interactivity somewhere

Agreed. Without making this transition seamless, SSR is pretty useless right now. This is a must-have to compete with the competitors in the market.

**Comment by andersme** [comment]
> > This one hurts pretty badly with .NET 8 if you have SSR with any kind of interactivity somewhere
> 
> Agreed. Without making this transition seamless, SSR is pretty useless right now. This is a must-have to compete with the competitors in the market.

And now I've simply replaced the WebAssembly interactivity with HTMX. What a cobbled mess Blazor has become.

**Comment by cbordeman** [comment]
This has been such a huge issue to so long, it kinds of blows my mind that the ASP.Net team keeps pushing it out as if it weren't important.

I agree with other commenters, after doing several Blazor sites in .Net 6, Blazor has become a cobbled mess.  I switched to Blazor Server to get around WASM load speed issues, but now I wish I'd pulled the trigger on HTMX several months ago.

**Comment by fingers10** [comment]
@javiercn I'm curious to know the plans to work on this issue. It's been there for a very long time from since I started using blazor WASM from the initial release. What's the thoughts of ASP.NET team on this?

**Comment by fingers10** [comment]
> > > This one hurts pretty badly with .NET 8 if you have SSR with any kind of interactivity somewhere
> > 
> > Agreed. Without making this transition seamless, SSR is pretty useless right now. This is a must-have to compete with the competitors in the market.
> 
> And now I've simply replaced the WebAssembly interactivity with HTMX. What a cobbled mess Blazor has become.

@andersme  you mean you switched to plain HTML? Can you explain more? I'm also looking for alternatives now.

**Comment by MarkusRodler** [comment]
HTMX is another competitor and is similar to Knockout.js and Angular 1. If someone wants to build websites like it's 2010, go for it 😄

**Comment by msynk** [comment]
Without this feature, preventing the flicker (or reducing its effects) requires a lot of work and creativity on both the server and the client side. Since other competitors have this (smooth hydration) out of the box, it has a big impact on projects developed by Blazor.

We love Blazor and the vast opportunities it opens for .NET developers, and I hope this won't be delayed any further than the .NET 10 milestone 🤞.

**Comment by jirisykora83** [comment]
> > > > This one hurts pretty badly with .NET 8 if you have SSR with any kind of interactivity somewhere
> > > 
> > > 
> > > Agreed. Without making this transition seamless, SSR is pretty useless right now. This is a must-have to compete with the competitors in the market.
> > 
> > 
> > And now I've simply replaced the WebAssembly interactivity with HTMX. What a cobbled mess Blazor has become.
> 
> [@andersme](https://github.com/andersme) you mean you switched to plain HTML? Can you explain more? I'm also looking for alternatives now.

basically, similar to https://alpinejs.dev/ it is fine when you need just very little interactivity and basically use Blazor as static rendering = fast loading time, no flickering etc..

but it is definitely not good tradeoff for anything other than very simple website (in terms of interactivity / front-end logic).

**Comment by oluatte** [comment]
Bummed to see the recent comments and current state on this issue. We also removed Blazor based interactivity from our app and mvoed to one of the HTMX variants (Datastar).

**Comment by mip1983** [comment]
I've not found this to be much of an issue myself. I'd rather not hold up the server side pre-render with async database type work, just to do it again when the page goes interactive. I'd rather that initial render was as fast as possible with loading indicators in place of interactive content, which is pretty straightforward.

Either, move your async code from OnInitializedAsync to OnAfterRenderAsync inside the first render if block, or wrap it in 'if (RendererInfo.IsInteractive). Voila, your page pre-renders fast, you do your async work once, and the content appears when it's interactive (no flicker) which is pretty quick when published.

If you're determined to have this stuff in the pre-render, have a look at the community stand up 'what's new for Blazor in .Net 10' on YouTube. About 40 mins in, you can see what you can do to persist state today, and what's coming. That's all overhead though, I think you'd need a pretty good reason to want to for pre-rendering. Makes a bit more sense for Blazor server and persisting state on reconnect.

**Comment by msynk** [comment]
@mip1983 It's not that simple. For example, consider the SEO of dynamic data, which is a big requirement. There are other important requirements that need the prerender to work, as explained in this issue.

**Comment by msynk** [comment]
Like I said, it's not that simple.
no workaround can completely resolve this issue without fixing this `completely replacing` the DOM while hydration.

**Comment by 3hxx** [comment]
Being able to pre-render drastically improves any SEO performance as it will send a skeleton of the HTML with all of your SEO friendly wording in it to the search engine.

**Comment by mip1983** [comment]
> Like I said, it's not that simple. no workaround can completely resolve this issue without fixing this `completely replacing` the DOM while hydrating.

What's an example of something you can't work around at the moment? Can you not pre-render and hydrate your SEO related content this way? https://www.youtube.com/live/IXU3hbnaX50?t=2218s

**Comment by msynk** [comment]
> What's an example of something you can't work around at the moment? Can you not pre-render and hydrate your SEO related content this way? https://www.youtube.com/live/IXU3hbnaX50?t=2218s

As I've said, it's not that simple, so I hope I can explain it clearly.
The problem discussed in this issue is not about the simple flickering shown in the video that could easily be solved with the `PersistentState` which we all are already using for the prerendering of our apps.

The problem, as explained in the issue itself, is something that happens at the lower levels of the Blazor rendering while hydrating the app after initialization for the first time when receiving the pre-rendered content.
The Blazor runtime, unlike other famous frameworks out there, instead of using the already rendered content of the HTML received from the server and just attaching to it, tries to re-rerender the whole UI and replace the already rendered DOM. This approach's problem is not that noticeable on simple UIs like the one shown in the video, but in real-world applications, this is noticeable and sometimes problematic.

**Comment by 3hxx** [comment]
What problems are your users experiencing? Do you have a video you can share?

**Comment by ysmoradi** [comment]
When a prerendered document is retrieved by the browser, CSS files begin downloading.
Once the CSS files are fully downloaded, the content becomes visible to the user.
At this point, animations start playing, users can scroll through nested items in lists, CSS-based carousels begin cycling to the next photos, and users can play any embedded videos.


While these user interface interactions are occurring, JavaScript files start downloading in the background, followed by hundreds of (WASM) files. After that, a prerendered state (encoded as a base64 JSON string located at the bottom of the document's body) is read and deserialized into the corresponding classes (The ASP.NET Core team has ensured that no System.Text source generators can be used in this process, deliberately 😅 keeping it slow)
A new DOM is then generated in memory and eventually replaces the existing DOM.

The process of downloading and executing code may take several seconds, depending on the client's network conditions, the size of the web app, and the processing power of the client's device CPU.

When this finishes, the list of items the user had scrolled through resets to the top, any video restarts from the beginning, animations replay, and the CSS-based carousel returns to the first photo. This moment is marked as the Largest Contentful Paint (LCP), even though the user could already read the content seconds earlier!

Why would you need a video to explain such an obvious scenario?

**Comment by oluatte** [comment]
Yeah, I don't understand the recent questions asking to justify this issue. The original issue was filed by an MS employee, the description in it is clear.  All the reactions above clearly indicate it is not a solitary problem.  I and others have ditched Blazor interactivity or Blazor entirely because of it. 

What is the skepticism here?

**Comment by mip1983** [comment]
I'm not saying this isn't a nice thing to get solved, but for people to say it makes Blazor interactivity unusable or that you have to jump to another framework seems a bit of an over the top reaction. I'm just trying to inject some first hand experience/realism and share some work arounds and upcoming changes that will help.

I have a my own real world Blazor site infront of me. It pre-renders quickly, you get the main layout, menu, placeholder widgets and loading indicators, there's a video background splash screen on load, CSS animations into a complex interactive (wasm) dashboard with charts and carousels, all of which play once with no flicker, going from 0 to interactive in ~1s (regardless of whether I'm on my desktop or low powered mobile). If you where to read ysmoradi description, you'd think that was basically impossible or really hard to do. It isn't.

**Comment by ysmoradi** [comment]
Let's talk with real examples. https://bitplatform.dev/ is the smallest website we've ever built using Blazor WebAssembly, with a total size of just 2.5MB. Another example is  https://sales.bitplatform.dev/, which comes in at 3MB. 

On a low to mid-range Android device with a 4G connection, these ultra-light websites take around 4 seconds to become interactive.  

I'm guessing you're referring to a desktop setup with a powerful PC and a high-speed network. If you share your website's URL here, I can say how long it would take to become interactive on a Galaxy A35 with a 4G connection @mip1983

**Comment by oluatte** [comment]
@mip1983 

Good for you. And your site sounds really fleshed out which is also great. I get that you're not seeing this issue but that fact by itself doesn't invalidate that others have seen this issue. I've been where you are (not seeing the issue others are seeing), and likely others have too. Never makes sense to dismiss other's experiences,  scenarios,  sense of urgency etc.  

Hydrating pre-rendered content is something nearly every FE framework does today. It's frankly embarrassing for something like this to be status quo that requires significant workarounds to implement (assuming it can be worked around).  

if you have helpful tips here or know of a given solution, we're all ears. Otherwise it's just a shite attitude.

**Comment by ysmoradi** [comment]
Let's talk with real examples. https://bitplatform.dev/ is the smallest website we've ever built using Blazor WebAssembly, with a total size of just 2.5MB. Another example is  https://sales.bitplatform.dev/, which comes in at 3MB. 

On a low to mid-range Android device with a 4G connection, these ultra-light websites take around 4 seconds to become interactive.  

I'm guessing you're referring to a desktop setup with a powerful PC and a high-speed network. If you share your website's URL here, I can say how long it would take to become interactive on a Galaxy A35 with a 4G connection @mip1983

**Comment by GajaLab** [comment]
I am a cms and website developer. The biggest limitation of Blazor is on web development where it is mandatory to use prerender for obvious reasons related to seo. Making the user wait a few seconds (especially on mobile) to use the interactive buttons is unacceptable. So much so that on the sites, with the exception of pages that can be rendered without prerender, such as the cart, I had to replace the interactivity with htmx.
Blazor is great for applications without prerendering, but for everything else it is not usable interactivity by today's standards.

**Comment by nw5428** [comment]
This is also causing a head ache for me.  We recently switched to allow pre rendering in a web assembly app. We have some forms where users are used to very quickly entering in form input.  When blazor changes from the pre-rendered state to interactive state, the form input loses focus and anything you have typed is also lost.

Can easily reproduce by starting from the visual studio template and doing a simple EditForm.  Type into a text box very quickly and when it switches to interactive you lose focus and if you were really fast what you've typed in is now gone.

My users were so annoyed that I had to revert to showing a loading spinner until RendererInfo.IsInteractive.  My users also hate this, but not has bad has having to potentially click twice to focus.  If blazor didn't replace the whole DOM I believe this would resolve this issue and result in a seamless transition.

**Comment by vladimir-angelov-1337** [comment]
The fact that this is still unsolved in 2025 is the reason why we will be ditching Blazor and switching to React, despite how much I love it.

Also calling this "technical debt" and not treating it as an absolutely essential feature of a SPA framework is insane.

**Comment by NearNihil** [comment]
I've not really done much with animations before, but now I am and it's *really* noticeable. Is it so much to ask to fix this? It's been years and it's causing headaches all round.

**Comment by ice6** [comment]
this issue is good showcase about github copilot's  inability. :)

**Comment by Ali1Jamal** [comment]
also moving from static page to interactive page or vice versa, when the theme of the browser or os is different from the website or app theme ( i add pwa ability to my blazor web app ), causes this flashing because of  replacing Dom and reload all assets.

https://github.com/user-attachments/assets/bf38beb6-39ea-4b75-a312-86b403f4d920

**Comment by BrendanRidenour** [comment]
Adding diagnostic evidence from .NET 10 with InteractiveWebAssembly that may be useful for anyone investigating this.

We confirmed that static HTML elements in App.razor (outside any component boundary) are destroyed and recreated during the SSR-to-WASM handoff. Specifically, a `<link rel="stylesheet">` that is plain markup in `<head>`, not rendered by any Blazor component.

**How we confirmed it:** We set all three DOM breakpoints (subtree modified, node removed, attribute modified) on the `<link>` element in Edge DevTools before triggering the navigation. After the WASM handoff completed, the breakpoints were gone without ever firing. This indicates the node was replaced as part of a wholesale parent replacement rather than individual node operations, which bypasses per-node mutation observers.

This means the wholesale DOM replacement described in the original issue extends beyond Blazor-managed component subtrees to static HTML that Blazor does not own.

**Workarounds evaluated:**

- `data-permanent`: only applies to enhanced navigation, not render mode transitions.
- `<link rel="preload">`: resource stays cached, but the FOUC still occurs because the DOM node itself is destroyed and recreated.
- Inline `<style>` fallback: blocked by Content Security Policy in our environment.
- Element ordering in `<head>`: no effect.

We have no remaining workaround available to us.

**Comment by cbordeman** [comment]
This is why everyone says Blazor is garbage now.  It's nearly unusable.

**Comment by pr8x** [comment]
.NET 11 is coming up and this is nowhere to be found on the roadmap.
